src/coreclr/jit/morph.cpp

```diff
@@ -11541,6 +11541,37 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
             }
             break;
 
+        case GT_IND:
+        {
+            // Fold IND(LDELEMA_REF(obj, index, type)) to INDEX(obj, index) if type is exact
+            if (op1->IsCall())
+            {
+                GenTreeCall* call = op1->AsCall();
+                if ((call->gtCallType == CT_HELPER) &&
+                    (eeGetHelperNum(call->gtCallMethHnd) == CORINFO_HELP_LDELEMA_REF))
+                {
+                    GenTree*             objOp   = call->gtCallArgs->GetNode();
+                    GenTree*             indexOp = call->gtCallArgs->GetNext()->GetNode();
+                    GenTree*             clsOp   = call->gtCallArgs->GetNext()->GetNext()->GetNode();
+                    CORINFO_CLASS_HANDLE cls     = gtGetHelperArgClassHandle(clsOp);
+                    if ((cls != NO_CLASS_HANDLE) && impIsClassExact(cls))
```

Comment:
Doesn't this need to check if the array element type is exact and equal to `cls` as well?

Reply:
Yes, not sure it's possible that they won't match in pure C#, but sure, I'll add

Reply:
Isn't it necessary to handle covariance, even for C#?

Reply:
> Isn't it necessary to handle covariance, even for C#?

not sure I understand what you mean - do you have an example where `ldelema's type is X (and it's sealed)` but the actual array is something different? (in C#)

Reply:
Ah, I misunderstood what `impIsClassExact` was checking, looks fine to me now.

Reply:
well, it worth checking types anyway so thanks for the suggestion 🙂